### PR TITLE
Fix S3242: Rule should not suggest IEnumerable<T> when there are multiple iterations over the collection

### LINKED ISSUE #680: Fix S3242: Rule should not suggest IEnumerable<T> when there are multiple iterations over the collection
### Description

Collection parameters in method signatures that are more derived than `IEnumerable<T>`, even when used only as enumerables, make sense. Specifically, multiple enumerations of an IEnumerable may have undesired side effects. Resharper even has a specific analyzer to catch that.

### Repro steps

The following method will trigger S3242, suggesting the use of `IEnumerable<int>` instead of `ICollection<int>`.
```cs
public static void Foo(ICollection<int> collection)
{
    foreach (var item in collection) { Console.WriteLine(item); }
    foreach (var item in collection.OrderBy(x => x)) { Console.WriteLine(item); }
}
```

### Expected behavior

S3242 should not fire because the method enumerates the collection twice and doing so on an IEnumerable can have undesired side effects.

### Actual behavior

S3242 is fired, suggesting a potentially dangerous fix.

### Related information

* SonarC# Version: SonarAnalyzer.CSharp 6.3.0.2862
* Visual Studio Version: 2017 (15.2)

### PR DESCRIPTION
**Pull request by michalb-sonar**
Fix #680

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs [hunk 1]
```diff
@@ -179,10 +179,10 @@ private ITypeSymbol FindParameterUseAsType(IdentifierNameSyntax identifier, Sema
             {
                 return semanticModel.GetTypeInfo(identifier).ConvertedType;
             }
-            else if (identifierParent is ElementAccessExpressionSyntax)
-            {
+            else if (identifierParent is ElementAccessExpressionSyntax)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs]: Same question you asked. What changed here?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs [hunk 2]
```diff
@@ -242,23 +242,23 @@ private INamedTypeSymbol FindOriginatingSymbol(ISymbol accessedMember)
...
-        private bool IsEventHandlerImplementation(IMethodSymbol methodSymbol)
-        {
-            // Cannot find a way to precisely tell whether this method is a event handler implementation
-            // without doing too much browsing in all trees.
-            // Let's go for a simple heuristic.
-            return methodSymbol.ReturnsVoid &&
-                methodSymbol.Parameters.Length == 2 &&
-                methodSymbol.Parameters[0].Name == "sender" &&
-                methodSymbol.Parameters[0].Type.Is(KnownType.System_Object);
+        private bool IsEventHandlerImplementation(IMethodSymbol methodSymbol)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs]: What changed here?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs [hunk 3]
```diff
@@ -309,6 +316,15 @@ private ISymbol FindMostGeneralType()
             {
                 var mostGeneralType = parameterSymbol.Type;
 
+                var multipleEnumerableCalls = usedAs.Where(kvp => kvp.Value > 1 &&
+                        kvp.Key.OriginalDefinition.Is(KnownType.System_Collections_Generic_IEnumerable_T)).ToArray();
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs]: Isn't this 2 tabs?

Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs]: Also don't we also need to check for the non-generic `IEnumerable`?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs (hunk 1 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs]: As we discussed, let's leave it for now (in both PRs) and fix properly - perhaps you'll be more successful with gitattributes...

**Comment by michalb-sonar** [comment]
@Evangelink one of the two classes should not be there...

**Pull request merged**